modules/tinymce/src/core/main/ts/keyboard/Nbsps.ts

```diff
@@ -149,22 +162,37 @@ const normalizeNbspInMiddleOfTextNode = (node: Text): boolean => {
   }
 };
 
-const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text): boolean => {
+const normalizeNbspAtEnd = (root: SugarElement<Node>, node: Text, makeNbsb: boolean): boolean => {
   const text = node.data;
   const lastPos = CaretPosition(node, text.length - 1);
-  if (isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos)) {
+  if (isNbspAt(text, text.length - 1) && !needsToBeNbsp(root, lastPos) && !makeNbsb) {
     node.data = text.slice(0, -1) + ' ';
     return true;
+  } else if (makeNbsb && isWhiteSpaceAt(text, text.length - 1) && needsToBeNbspRight(root, lastPos)) {
+    node.data = text.slice(0, -1) + Unicode.nbsp;
+    return true;
   } else {
     return false;
   }
 };
 
-const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> => Optional.some(pos).filter(hasNbsp).bind((pos) => {
-  const container = pos.container() as Text;
-  const normalized = normalizeNbspAtStart(root, container) || normalizeNbspInMiddleOfTextNode(container) || normalizeNbspAtEnd(root, container);
-  return normalized ? Optional.some(pos) : Optional.none();
-});
+const normalizeNbsps = (root: SugarElement<Node>, pos: CaretPosition): Optional<CaretPosition> =>
+  Optional.some(pos).filter(hasNbsp).fold(
+    () => {
+      if (needsToBeNbsp(root, pos) && SugarNode.isText(SugarElement.fromDom(pos.container()))) {
```

Comment:
This should be using `NodeType.isText()` instead of converting it to a SugarElement just to check this.

Reply:
That's much neater.